File: core/src/main/java/com/devexperts/dxlab/lincheck/ActorGenerator.java
package com.devexperts.dxlab.lincheck;

/*
 * #%L
 * core
 * %%
 * Copyright (C) 2015 - 2017 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import java.lang.reflect.Method;
import java.util.List;
import java.util.stream.Collectors;

/**
 * The instances of this class are used to generate {@link Actor actors}
 * using {@link ParameterGenerator parameter generators}.
 */
class ActorGenerator {
    private final Method method;
    private final List<ParameterGenerator<?>> parameterGenerators;
    private final List<Class<? extends Throwable>> handledExceptions;
    private final boolean useOnce;

    ActorGenerator(Method method, List<ParameterGenerator<?>> parameterGenerators,
        List<Class<? extends Throwable>> handledExceptions, boolean useOnce)
    {
        this.method = method;
        this.parameterGenerators = parameterGenerators;
        this.handledExceptions = handledExceptions;
        this.useOnce = useOnce;
    }

    Actor generate() {
        return new Actor(method, parameterGenerators.stream()
            .map(ParameterGenerator::generate).collect(Collectors.toList()), handledExceptions);
    }

    public boolean useOnce() {
        return useOnce;
    }
}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ActorGenerator.kt
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package org.jetbrains.kotlinx.lincheck.execution

import org.jetbrains.kotlinx.lincheck.*
import org.jetbrains.lincheck.datastructures.ParameterGenerator
import org.jetbrains.lincheck.datastructures.THREAD_ID_TOKEN
import java.lang.reflect.*
import java.util.Random

/**
 * Implementations of this class generate [actors][Actor]
 * using [parameter generators][org.jetbrains.lincheck.datastructures.ParameterGenerator].
 */
class ActorGenerator(
    private val method: Method,
    private val parameterGenerators: List<ParameterGenerator<*>>,
    val useOnce: Boolean,
    cancellableOnSuspension: Boolean,
    private val blocking: Boolean,
    private val causesBlocking: Boolean,
    promptCancellation: Boolean
) {
    private val cancellableOnSuspension = cancellableOnSuspension && isSuspendable
    private val promptCancellation = cancellableOnSuspension && promptCancellation

    fun generate(threadId: Int, random: Random): Actor {
        val parameters = parameterGenerators
            .map { it.generate() }
            .map { if (it === THREAD_ID_TOKEN) threadId else it }
        val cancelOnSuspension = this.cancellableOnSuspension and random.nextBoolean()
        val promptCancellation = cancelOnSuspension and this.promptCancellation and random.nextBoolean()
        return Actor(
            method = method,
            arguments = parameters,
            cancelOnSuspension = cancelOnSuspension,
            blocking = blocking,
            causesBlocking = causesBlocking,
            promptCancellation = promptCancellation
        )
    }

    val isSuspendable: Boolean get() = method.isSuspendable()
    override fun toString() = method.toString()
}